guppylang-internals/src/guppylang_internals/std/_internal/checker.py

```diff
@@ -502,7 +504,28 @@ def synthesize(self, args: list[ast.expr]) -> tuple[ast.expr, Type]:
 
 
 class WasmCallChecker(CustomCallChecker):
+    wasm_sigs: ConcreteWasmModule
+
+    def setup(self, wasm_sigs: ConcreteWasmModule) -> None:
+        self.wasm_sigs = wasm_sigs
+
     def check(self, args: list[ast.expr], ty: Type) -> tuple[ast.expr, Subst]:
+        type_in_wasm: FunctionType = DEF_STORE.wasm_functions[self.func.id]
+
+        # Drop the first arg because it should be "self"
+        expected_type = FunctionType(self.func.ty.inputs[1:], self.func.ty.output)
+
+        if expected_type != type_in_wasm:
```

Comment:
Maybe this check would make more sense in `RawWasmFunctionDef.parse` instead of checking every time we call the function?